### PR TITLE
The first objective for M2D Visualization

### PR DESCRIPTION
**Pull request by jl-liu2022**
History:

This change allows one to track a specific real packet during a simulation. This is a PR of issue #1947.

**Description**

* Change the type of input "track_rpacket" from boolean to number.
* A new temporary variable "tracking" is input to "single_packet_loop" to decide whether to track.

**Motivation and context**

The mission "Monte Carlo 2D Packet Visualization" in idea list of TARDIS in gsoc 2022

**How has this been tested?**
- [ ] Testing pipeline.
- [x] Other. 
I try to test with "python setup.py test", but it told me to use "tox -e test". However, I find that the test fails even when no change is made. 
[Screen Shot 2022-04-12 at 22.17.28.pdf](https://github.com/tardis-sn/tardis/files/8474635/Screen.Shot.2022-04-12.at.22.17.28.pdf)
So I just test whether my code work correctly. See the link below. 

**Examples**

[file:///Users/pro/TARDIS/one_packet_tracking/one_packet_tracking.html](url)

**Type of change**

- [ ] Bug fix. 
- [x] New feature. 
- [ ] Breaking change. 
- [ ] None of the above. 

**Checklist**

- [x] My change requires a change to the documentation.
    - [ ] I have updated the documentation accordingly.
    - [ ] (optional) I have built the documentation on my fork following [the instructions](https://tardis-sn.github.io/tardis/contributing/development/documentation_guidelines.html#sharing-the-built-documentation-in-your-pr-documentation-preview).
- [ ] I have assigned and requested two reviewers for this pull request.

Now:
Undo the changes and just add a py file for the first objective. 

Add shell boundaries.

Here is the pdf of the notebook: 
[one_packet_tracking.pdf](https://github.com/tardis-sn/tardis/files/8513337/one_packet_tracking.pdf)

Here is the figure of mu vs. r:
plot_mu_r

### CODE REVIEW

File: tardis/io/schemas/montecarlo.yml [hunk 1]
```diff
@@ -69,8 +69,8 @@ properties:
     default: {}
     properties:
       track_rpacket:
-        type: boolean
```
Comment (Rodot-) [inline tardis/io/schemas/montecarlo.yml:72]: Please remove all changes from the schema files.  Tracking the r-packets is something that is set by the user in their own config file.

File: tardis/montecarlo/montecarlo_configuration.py [hunk 2]
```diff
@@ -15,5 +15,5 @@
 INITIAL_TRACKING_ARRAY_LENGTH = None
 LEGACY_MODE_ENABLED = False
 VPACKET_LOGGING = False
-RPACKET_TRACKING = False
```
Comment (Rodot-) [inline tardis/montecarlo/montecarlo_configuration.py:18]: See above

File: tardis/montecarlo/montecarlo_numba/base.py [hunk 3]
```diff
@@ -227,13 +227,23 @@ def montecarlo_main_loop(
         vpacket_collection = vpacket_collections[i]
         rpacket_tracker = rpacket_trackers[i]
 
+        if montecarlo_configuration.RPACKET_TRACKING == 0:
```
Comment (Rodot-) [inline tardis/montecarlo/montecarlo_numba/base.py:230]: This project should be done purely externally from the internal workings of tardis.  Please do not change the default behavior if it is not part of the project.

**Review by Rodot- — COMMENTED** [review]
Please carefully read the objective for this project.  There is no requirement to modify the internal code of tardis to complete this objective.  Everything should be done from external code which runs tardis and generates the plots

File: first_objective_of_M2DVisualization.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
+'''This is a code to realize the first project of M2D visualization, running a simulation with logging enabled with only one packet.'''
+from tardis import run_tardis
+from tardis.io.config_reader import Configuration
+from tardis.io.atom_data.util import download_atom_data
+
+# Downloading atom data
+download_atom_data('kurucz_cd23_chianti_H_He')
+
+# Reading the Configuration stored in `tardis_config_packet_tracking.yml` into␣ ↪config
+config = Configuration.from_yaml("tardis_example.yml")
```
Comment (jaladh-singhal) [inline first_objective_of_M2DVisualization.py:10]: Don't add this file on root directory if you haven't modified it. The example config already exists in `docs/` dir - use that

File: first_objective_of_M2DVisualization.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+sim = run_tardis(config, show_convergence_plots=False, show_progress_bars=False)
+
+# Plot mu vs. radius
+from matplotlib import pyplot as plt
+
+fig, ax = plt.subplots()
+ax.set_xlabel('r [cm]')
+ax.set_ylabel('$\\rm \\mu$')
+ax.plot(sim.runner.rpacket_tracker[0].r, sim.runner.rpacket_tracker[0].mu) 
+plt.show()
```
Comment (jaladh-singhal) [inline first_objective_of_M2DVisualization.py:28]: You can also show shell boundaries in this plot

**Review by jaladh-singhal — COMMENTED** [review]
@jl-liu2022 Thanks for this PR. Can you please also post image of the plot in PR description?

**Comment by jl-liu2022** [comment]
> @jl-liu2022 Thanks for this PR. Can you please also post image of the plot in PR description?

Yes, of course.

File: first_objective_of_M2DVisualization.py (hunk 5 above)
Comment (jl-liu2022) [reply first_objective_of_M2DVisualization.py:43]: I have added the shell boundaries. I also want to show some information of  interactions, but it seems that only the last interaction types are recorded.

Comment (jl-liu2022) [reply first_objective_of_M2DVisualization.py:43]: > You can also show shell boundaries in this plot

I have added the shell boundaries. I also want to show some information of  interactions, but it seems that only the last interaction types are recorded.

File: tardis/montecarlo/montecarlo_numba/base.py [hunk 6]
```diff
@@ -226,7 +226,7 @@ def montecarlo_main_loop(
 
         vpacket_collection = vpacket_collections[i]
         rpacket_tracker = rpacket_trackers[i]
-
```
Comment (Rodot-) [inline tardis/montecarlo/montecarlo_numba/base.py:229]: Can you please remove this change from the commit history?

**Comment by Rodot-** [comment]
This simulation in your notebook seems to be taking a while to run for a single packet, can you set the number of iterations to 1?

File: tardis/montecarlo/montecarlo_numba/base.py (hunk 6 above)
Comment (jl-liu2022) [reply tardis/montecarlo/montecarlo_numba/base.py]: Sure.

**Pull request closed without merging**